### PR TITLE
Allow adapter to be passed in via options

### PR DESCRIPTION
**Pull request by nrkn**
- arbitrary backing dom structures via `options.adapter`
- no breaking changes, external API is the same
- uses domutils by default
- static methods on CSSselect always use domutils
- includes browser-adapter for browser DOM
- tests
- readme updated

### CODE REVIEW

**Comment by nrkn** [comment]
While I'm fixing my numerous eslint issues, I notice that there is rule failing that I didn't introduce, the indented multiline `var`, eg:

```javascript
var DomUtils       = require("domutils"),
    falseFunc      = require("boolbase").falseFunc,
    compileFactory = require("./lib/compile.js"),
    browserAdapter = require("./browser-adapter"),
    defaultCompile = compileFactory(DomUtils);
```

Would you like me to change these to multiple `var` declarations instead while I'm fixing the things I've introduced, or are you planning to change the `eslint` rules and would you prefer I leave them?

File: browser-adapter.js [hunk 1]
```diff
@@ -51,17 +90,15 @@ module.exports = {
 		}
 		return result;
 	},
-	//https://github.com/ded/qwery/blob/master/pseudos/qwery-pseudos.js#L47-54
 	getText: function getText(elem) {
-		var str = "",
-		    childs = getChildren(elem);
```
Comment (fb55) [inline browser-adapter.js]: `adapter.getText` -> `getText`

File: lib/attributes.js [hunk 2]
```diff
@@ -1,181 +1,186 @@
...
-var attributeRules = {
-	__proto__: null,
-	equals: function(next, data){
-		var name  = data.name,
-		    value = data.value;
-
-		if(data.ignoreCase){
-			value = value.toLowerCase();
-
-			return function equalsIC(elem){
```
Comment (fb55) [inline lib/attributes.js]: This should be unpacked in the closure.

File: index.js [hunk 3]
```diff
@@ -2,57 +2,78 @@
...
+		    compile,
+		    compileUnsafe;
+
+		if(adapter === DomUtils){
+			compile = defaultCompile;
+		} else {
+			compile = compileFactory(adapter);
+		}
+
+		compileUnsafe = compile.compileUnsafe;
```
Comment (fb55) [inline index.js]: Not sure I can follow why separate functions are required. I would have approached this by

1. Making sure `options` is always an object (`options = options || {}`)
2. And ensuring its `adapter` property is set (`options.adapter = options.adapter || DomUtils`)

File: test/adapter.js [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+		assert.equal(pseudo.length, 2);
+	});
+
+	it("should :first-child", function(){
+		var body = getBody(html);
+		var pseudo = select("div:first-child", body, options);
+
+		assert.equal(pseudo.length, 1);
+	});
+});
```
Comment (fb55) [inline test/adapter.js]: This is awesome. Perhaps it would be an idea to have the browser adapter as a separate module, as it is starting to look like much more than just an experiment.

File: README.md [hunk 5]
```diff
@@ -86,6 +87,73 @@ Arguments are the same as for `CSSselect(query, elems)`. Only returns the first
...
+```
+
+The method signature notation used below should be fairly intuitive - if not,
+see the [`rtype`](https://github.com/ericelliott/rtype) or
+[`TypeScript`](https://www.typescriptlang.org/) docs, as it is very similar to
+both of those. You may also want to look at
+[`domutils`](https://github.com/fb55/domutils) to see the default
+implementation.
+
+```ts
```
Comment (fb55) [inline README.md]: Providing an API specification is a great idea. It might be useful to have this as an actual TypeScript or Flowtype definition.

Comment (nrkn) [inline README.md]: Something like this?

https://gist.github.com/nrkn/b0b096ec4d48b2c590af7ec2301ae5e3

File: index.js (hunk 3 above)
Comment (nrkn) [inline index.js]: Do you mean simply changing to something like this:

```javascript
function getSelectorFunc(searchFunc){
	return function select(query, elems, options){
		options = options || {};
		options.adapter = options.adapter || DomUtils;

		var adapter = options.adapter,
		    compile,
		    compileUnsafe;

		if(adapter === DomUtils){
			compile = defaultCompile;
		} else {
			compile = compileFactory(adapter);
		}

		compileUnsafe = compile.compileUnsafe;

		// etc.
}
```

Or were you asking about why I'm assigning the `compile` variable like that etc.? It's so that the static methods on CSSselect are always using the same instance of compile with psuedos and filters attached so that it doesn't break when the pseduos are patched at runtime by `compile.js`

Or did I misunderstand this comment entirely, haha?

File: test/adapter.js (hunk 4 above)
Comment (nrkn) [inline test/adapter.js]: Yeah definitely, that also gets rid of having jsdom as a dev dependency, it makes the tests slow. Shall I remove it and if so do you want to publish it or shall I?

Comment (fb55) [inline test/adapter.js]: Feel free to publish it (& link to the package in the readme)!

**Comment by nrkn** [comment]
OK regards moving browser-adapter, it's here, I'll publish when the pull is resolved:

https://github.com/nrkn/css-select-browser-adapter

Regards passing options around instead, I started doing it that way, but it 
meant modifying code *all* over the place and I kept breaking things by adding 
option args to various functions where DomUtils would have been in scope but 
that weren't designed to take options (including pseudos, I had issues with 
the `verifyArgs` stuff and it gave me a headache trying to figure out why) so I 
looked for a simpler solution and felt that leaving the majority of the code 
basically unchanged and making the adapter available through a closure was 
simpler and tidier and less likely to have unintended side effects that I 
couldn't predict - basically just changing the way that some modules export 
their contents so that adapter was passed in rather than DomUtils being 
required. 

Please let me know if I'm fundamentally misunderstanding something here!

**Comment by fb55** [comment]
@nrkn Sorry for dropping the ball on this. For now, the current approach is okay, please rebase and I'll merge.

**Comment by nrkn** [comment]
@fb55 I think that's it!

**Comment by fb55** [comment]
@nrkn I decided to merge and clean it up afterwards. Thanks a lot for this PR!
